src/User/Access/Gate.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace Flarum\User\Access;
+
+use Flarum\Database\AbstractModel;
+use Flarum\Event\GetPermission;
+use Flarum\User\User;
+use Illuminate\Contracts\Events\Dispatcher;
+use Illuminate\Support\Arr;
+
+class Gate
+{
+    protected const EVALUATION_CRITERIA = [
+        AbstractPolicy::FORCE_DENY => false,
+        AbstractPolicy::FORCE_ALLOW => true,
+        AbstractPolicy::DENY => false,
+        AbstractPolicy::ALLOW => true,
+    ];
+
+    /**
+     * @var Dispatcher
+     */
+    protected $events;
+
+    /**
+     * @var array
+     */
+    protected $policies;
+
+    /**
+     * @param Dispatcher $events
+     */
+    public function __construct(Dispatcher $events, array $policies)
+    {
+        $this->events = $events;
+        $this->policies = $policies;
+    }
+
+    /**
+     * Determine if the given ability should be granted for the current user.
+     *
+     * @param  User $actor
+     * @param  string  $ability
+     * @param  string|AbstractModel $model
+     * @return bool
+     */
+    public function allows(User $actor, string $ability, $model): bool
+    {
+        $results = [];
+        $appliedPolicies = [];
+
+        if ($model) {
+            $modelClasses = is_string($model) ? [$model] : array_merge(class_parents(($model)), [get_class($model)]);
+
+            foreach ($modelClasses as $class) {
+                $appliedPolicies = array_merge($appliedPolicies, Arr::get($this->policies, $class, []));
+            }
+        } else {
+            $appliedPolicies = Arr::get($this->policies, AbstractPolicy::GLOBAL);
+        }
+
+        foreach ($appliedPolicies as $policy) {
+            $results[] = $policy->checkAbility($actor, $ability, $model);
```

Comment:
I still believe policy classes should only be resolved at this time to save on resources, skip resolving unnecessary constructor dependencies and delay resolves as much as possible.

Reply:
Hmm, it was moved here to stop resolving them multiple times. Maybe we can resolve them here the first time needed and store that?